Create ivtools

cwhanse opened the pull request:
pvlib python pull request guidelines
====================================

Thank you for your contribution to pvlib python! You may delete all of these instructions except for the list below.

You may submit a pull request with your code at any stage of completion.

The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items below:

 - [x] Closes #511 
 - [x] I am familiar with the [contributing guidelines](http://pvlib-python.readthedocs.io/en/latest/contributing.html).
 - [x] Fully tested. Added and/or modified tests to ensure correct behavior for all reasonable inputs. Tests (usually) must pass on the TravisCI and Appveyor testing services.
 - [x] Updates entries to `docs/sphinx/source/api.rst` for API changes.
 - [x] Adds description and name entries in the appropriate `docs/sphinx/source/whatsnew` file for all changes.
 - [x] Code quality and style is sufficient. Passes LGTM and SticklerCI checks.
 - [x] New code is fully documented. Includes sphinx/numpydoc compliant docstrings and comments in the code where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.

Brief description of the problem and proposed solution (if not already fully described in the issue linked to above):

Add module ivtools with code to fit diode models and diode equations using IV curve data.

cwhanse commented:
Test failures for 2.7, 2.7min and 3.4, because the SAM library nrel-pysam is available from 3.5.  How do I mark the test to be skipped on lower python versions?  Otherwise, ready for review.

markcampanelli commented:
> Test failures for 2.7, 2.7min and 3.4, because the SAM library nrel-pysam is available from 3.5. How do I mark the test to be skipped on lower python versions? Otherwise, ready for review.

I hope to have https://github.com/pvlib/pvlib-python/pull/728 up for review today, which is intended to circumvent this very problem.

wholmgren commented:
What's the use case for the nrel-pysam wrapper function? It's not clear to me why someone would use the pvlib wrapper instead of directly calling the nrel-pysam function.

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,215 @@
+    # Find intercept and slope of linear portion of IV curve.
+    # Start with V < vlim * v_oc, extend by adding points until slope is
+    # acceptable
+    beta = [np.nan for i in range(5)]
```
What is beta? Why are there 5 elements?

cwhanse replied:
Coefficients in an equation to appear in the reference which is in preparation.

cwhanse replied:
I will write out the re-arranged single diode equation in the docstring Notes section.  That will show how `beta` relates to the five diode parameters.

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,215 @@
+                beta[0] = coef[1].item()
+                # sign change of slope to get positive parameter value
+                beta[1] = -coef[0].item()
+        except Exception as e:
```
So far as I can tell, these `try; except Exception as e; raise e` blocks don't serve any purpose.

cwhanse replied:
You're right.  Either provide a message that helps a pvlib user understand why the np.polyfit error comes about, or delete and let the np.polyfit error stand alone.

cwhanse replied:
`numpy.polyfit` raises `ValueError` or `TypeError` when inputs are wrong shape or empty.  I think that's sufficient.  `numpy.polyfit` raises `RankWarning` when the fit is suspect.

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,215 @@
+    beta = [np.nan for i in range(5)]
+    # get index of largest voltage less than/equal to limit
+    idx = _max_index(v, vlim * v_oc)
+    while np.isnan(beta[1]) and (idx <= len(v)):
```
Is this loop guaranteed to end? It's not clear to me.

cwhanse replied:
as long as `v` is finite, yes.  Could be more clear, I agree.  Would it be better to initialize a logical `found` and update to `found = True` when a good solution for `beta[0] and beta[1]` is obtained?

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,215 @@
+        except Exception as e:
+            raise e
+
+    if not any([np.isnan(beta[i]) for i in [0, 1, 3, 4]]):
```
I suggest reversing the blocks of the `if/else` so that you can remove the `not` part of the statement. I think that would be slightly more clear.

No need for the list comprehension brackets within `any` or `all` functions.

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,215 @@
+            idx += 1
+
+    if not np.isnan(beta[0]):
+        # Find parameters from exponential portion of IV curve
```
Each of the while and if blocks could call out to a (possibly private) function. I think it would improve the readability.

cwhanse replied:
Agree

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,215 @@
+
+def _max_index(x, xlim):
+    """ Finds maximum index of value of x <= xlim """
+    return int(np.argwhere(x <= xlim)[-1])
```
If I understand correctly, you're assuming that the input is sorted. If so, I think `np.searchsorted` does the job for you.

cwhanse replied:
Yes.  Thanks!

cwhanse commented:
> What's the use case for the nrel-pysam wrapper function? It's not clear to me why someone would use the pvlib wrapper instead of directly calling the nrel-pysam function.

Name translation and a testable wrapper, when the nrel-pysam functions are in flux.  When I started this PR the convenient `PySSC.ssc_sim_from_dict` wasn't available, and the interface required a quite a bit more work.  With `PySSC.ssc_sim_from_dict` the wrapper is mostly for connecting names.

I have in mind a suite of similar functions, `fit_cec_with_sam` is one, `fit_cec_snl` could be another that would derive from the PR @mikofski is working.

cwhanse commented:
stickler appears stuck.  otherwise I think this is ready.

wholmgren commented:
Can someone (maybe @mikofski @markcampanelli @adriesse ?) give this a review? Not my area of expertise.

I'm ok with the API. Minor fixes:

* The pip section of the CI requirements files duplicates items in the conda section. 
* `nrel-pysam` needs to be added to `setup.py` optional dependencies: https://github.com/pvlib/pvlib-python/blob/master/setup.py#L48

adriesse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+def _find_beta0_beta1(v, i, vlim, v_oc):
+    # Get intercept and slope of linear portion of IV curve.
+    # Start with V =< vlim * v_oc, extend by adding points until slope is
+    # acceptable
```
Briefly explain what is acceptable

adriesse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+    """
+
+    # Find beta0 and beta1 from linear portion of the IV curve
+    beta0, beta1 = _find_beta0_beta1(v, i, vlim, v_oc)
```
I would prefer nomenclature related to the task at hand rather than greek.

cwhanse replied:
So would I.  But the beta coefficients don't have ready interpretation as do the five SDE parameters.  Open to suggestions.

adriesse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+        (Rs*I)/(nNsVth) = beta2 + beta3*V + beta4*I
+
+    Values for ``IL, I0, Rs, Rsh,`` and ``nNsVth`` are calculated from the
+    regression coefficents beta0, beta1, beta3 and beta4.
```
`beta2` not needed?

cwhanse replied:
Not used.  beta2 is related to log(I0). In testing, we found that beta2 to have significant variance when measurement noise is present, and so calculating I0 from beta2 presented problems. There's an averaging technique in `_calculate_sde_parameters` to get I0 from the other four parameters which seems to be stable.

adriesse commented on pvlib/test/test_ivtools.py:
```diff
@@ -0,0 +1,88 @@
+    return {'V_mp_ref': 46.6, 'I_mp_ref': 4.73, 'V_oc_ref': 58.3,
+            'I_sc_ref': 5.05, 'alpha_isc': 0.000495, 'beta_voc': -0.003372,
+            'gamma_pmp': -0.43, 'cells_in_series': 96}
+
```
Just curious, why are these functions?

cwhanse replied:
To be re-used in tests for other fitting methods.

adriesse replied:
Still curious to understand what flexibility the function adds compared to a dict with the same name...

markcampanelli replied:
What's nice about the function is that it returns a copy of the dictionary that can be mutated without mutating the original dictionary. In testing, this can be useful for changing one or two parameters off of a base case.

adriesse replied:
Ok, instead of `dict.copy()`?  The luxury of always having multiple ways to achieve something in python!  :)

markcampanelli replied:
Inadvertently mutating dictionaries (or other objects) in Python remains one of my biggest challenges with the language. I admittedly wasn't sure of the behavior here until I ran some examples.

wholmgren replied:
When using `pytest`, it's preferable mark the setup function with `@pytest.fixture` and then pass the name of the setup function into the test function. `pytest` calls the fixture each time a new test is run (unless otherwise specified using the `scope` keyword argument). So this function becomes:

```
@pytest.fixture
def cec_params_cansol_cs5p_220p():
    return {'V_mp_ref': 46.6, 'I_mp_ref': 4.73, 'V_oc_ref': 58.3,
            'I_sc_ref': 5.05, 'alpha_sc': 0.0025, 'beta_voc': -0.19659,
            'gamma_pmp': -0.43, 'cells_in_series': 96}
```

and the test function becomes

```
def test_fit_cec_sam(cec_module_parameters, cec_params_cansol_cs5p_220p):
# no need to call get_cec_params_cansol_cs5p_220p()
```

cwhanse replied:
what is the returned dict assigned to, so it can be referenced?

wholmgren replied:
When `pytest` runs the test function, the fixture's returned dict is assigned to `cec_params_cansol_cs5p_220p` within the test function. `pytest` creates new returned dict if the fixture is used in another test function. https://docs.pytest.org/en/latest/fixture.html

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+import numpy as np
+
+
+def fit_cec_with_sam(celltype, v_mp, i_mp, v_oc, i_sc, alpha_sc, beta_voc,
```
I think the "with" is extraneous. It is not used in the sandia-based name.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+        Voltage at maximum power point at standard test condition (STC)
+    i_mp : float
+        Current at maximum power point at STC
+    Voc : float
```
Typo. Should be `v_oc`.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+    '''
+    Estimates parameters for the CEC single diode model using the SAM SDK.
+
+    Parameters
```
Specify units for each parameter, where applicable.

Perhaps document the allowable ranges of the input parameters.

cwhanse replied:
Added units

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+
+    Adjust : float
+        The adjustment to the temperature coefficient for short circuit
+        current, in percent
```
My understanding is that this adjustment is simultaneously applied to `beta_voc`.

cwhanse replied:
It is, but `beta_voc` is a model output, it is not used when computing an IV curve.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+        The shunt resistance at reference conditions, in ohms.
+
+    R_s : float
+        The series resistance at reference conditions, in ohms.
```
I would love to put the `_ref` suffix here too, but I realize there is precedent not to.

cwhanse replied:
I agree but am not including `_ref` to keep the SAM variable names. Page 41 of [NREL manual](https://www.nrel.gov/docs/fy15osti/64102.pdf)

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+    try:
+        from PySAM import PySSC
+    except ImportError as e:
+        raise(e)
```
I don't think the try-except is necessary.

cwhanse replied:
Perhaps not. I don't know how the code performs if PySAM isn't found. I think explicitly raising `ImportError` could help. One challenge, there is a python package `pysam` and then there's `NREL-PySAM`, it's the latter that is needed. Do you think that raising a specific message (rather than the `ImportError`) would help here?

markcampanelli replied:
It looks like a `ModuleNotFoundError` exception (not `ImportError`) is raised if the package is not installed:
```terminal
$ python -c "from PySAM import PySSC"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'PySAM'
```
Because of the naming similarity issue, it might be worth "raising further" from within the original exception with a message that points to NREL's package. This way you don't clobber the original exception's information, which might be useful for debugging the problem. Perhaps something like:
```python
try:
    from PySAM import PySSC
except Exception as e:
        raise ImportError("Make sure you have NREL's PySAM package installed correctly. It is found at https://pypi.org/project/NREL-PySAM/.") from e
```
See https://stackoverflow.com/questions/6062576/adding-information-to-an-exception

wholmgren replied:
I like @markcampanelli's suggestion with a minor change: `ModuleNotFoundError` is a subclass of `ImportError`, so `except ImportError` will work for any kind of import problem.

markcampanelli replied:
Sorry I missed the subclass relationship. I'm ever the skeptic, so I'm less comfortable limiting the scope of what could have gone wrong with the import. As long as the other exceptions don't get swallowed I'm fine with limiting the except clause to only `ImportError`s.

cwhanse replied:
Done. Thanks @markcampanelli

markcampanelli commented on setup.py:
```diff
@@ -45,7 +45,7 @@
 EXTRAS_REQUIRE = {
     'optional': ['ephem', 'cython', 'netcdf4', 'numba', 'pvfactors', 'scipy',
-                 'siphon', 'tables'],
+                 'siphon', 'tables', 'nrel-pysam'],
```
Could you maintain alphabetical ordering?

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+        Value is one of 'monoSi', 'multiSi', 'polySi', 'cis', 'cigs', 'cdte',
+        'amorphous'
+    v_mp : float
+        Voltage at maximum power point at standard test condition (STC)
```
Here and elsewhere. I think STC should more generally be called reference conditions (RC), which is used in the outputs. If I understand correctly, nothing about this fit requires that RC be STC.

cwhanse replied:
I think that change is a little risky. Unlike the reference temperature, SAM provides no keyword to specify the reference irradiance condition, and the [documentation](https://www.nrel.gov/docs/fy15osti/64102.pdf) page 43 shows 1000 W/m2 explicitly in the equations for `R_SH` and `I_L`. For this reason, I think the input values `v_mp` etc. have to be at 1000 W/m2 and the reference temperature

markcampanelli replied:
If the reference temperature is not the default 25 degC, then STC is a misnomer, correct?

cwhanse replied:
Yes, it is. I looked through the NREL source to confirm that 1) Tref is used and 2) no irradiance is assumed in the fitting.  Changing the docstring and adding a note about how the reference condition is implicitly specified.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+    a_ref = result['a']
+    I_L_ref = result['Il']
+    I_o_ref = result['Io']
+    R_s = result['Rs']
```
Again, the absent `_ref` always sticks out like a sore thumb to me.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+def fit_cec_with_sam(celltype, v_mp, i_mp, v_oc, i_sc, alpha_sc, beta_voc,
+                     gamma_pmp, cells_in_series, temp_ref=25):
+    '''
+    Estimates parameters for the CEC single diode model using the SAM SDK.
```
Perhaps reference "An Improved Coefficient Calculator for the CEC 6 Parameter Photovoltaic Module Model" or the SAM SDK documentation here?

cwhanse replied:
Added Dobos reference. The SAM SDK function 6parsolve isn't documented anywhere that I could find.

markcampanelli commented:
@cwhanse Fingers crossed, I should have time this evening to review the rest of this PR.

markcampanelli commented on pvlib/test/test_ivtools.py:
```diff
@@ -0,0 +1,88 @@
+import pytest
+from pvlib import pvsystem
+from pvlib import ivtools
+from conftest import requires_scipy, requires_pysam
```
I could not import this file into a Jupyter notebook (in order to get at the test data) without changing this line to
```python
from pvlib.test.conftest import requires_scipy, requires_pysam
```

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,268 @@
+                'beta_voc': beta_voc, 'gamma_pmp': gamma_pmp,
+                'Nser': cells_in_series, 'Tref': temp_ref}
+
+    result = PySSC.ssc_sim_from_dict(datadict)
```
Need to check for fitting success before assigning results, because dictionary entries are apparently missing altogether when it fails causing a `KeyError` at assignment:
```python
if result['cmod_success'] == 1:
    # Assign everything
else:
    # Assign nan's like in fit_sde_sandia()?
```
This behavior also needs to be tested and documented.

Here's how I broke it:
```python
v_mp = 0.45
i_mp = 5.25
v_oc = 0.55
i_sc = 5.5
alpha_sc = 0.0005 * i_sc
beta_voc = 0.005 * v_oc
gamma_pmp = 0.0055
cells_in_series = 1
temp_ref = 25
for celltype in ['monoSi', 'multiSi', 'polySi', 'cis', 'cigs', 'cdte', 'amorphous']:
    ivtools.fit_cec_sam(celltype, v_mp, i_mp, v_oc, i_sc, alpha_sc, beta_voc, gamma_pmp, cells_in_series, temp_ref)
```

cwhanse replied:
Thanks!

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,268 @@
+
+
+def fit_sde_sandia(v, i, v_oc, i_sc, v_mp, i_mp, vlim=0.2, ilim=0.1):
+    """ Fits the single diode equation to an IV curve.
```
I always like to say "Fits the single diode equation (SDE) to an I-V curve for a device assumed to be at constant temperature and irradiance."

cwhanse replied:
Added to the Notes section.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,268 @@
+        defines linear portion of IV curve i.e. V <= vlim * v_oc
+
+    ilim : float, default 0.1
+        defines exponential portion of IV curve i.e. I > ilim * i_sc
```
Model question: Is it possible to determine `ilim` from `vlim` (or vice versa)?

This would simplify the interface somewhat.

markcampanelli replied:
Ah, I suppose "exponential portion" excludes the knee of the curve. However, another interpretation of "exponential portion" is where the exponential term cannot be neglected. Is the sign correct in `I > ilim * i_sc`?

cwhanse replied:
The intent is the latter, where the exponential term is numerically meaningful. However, `ilim` protects against log(negative numbers) so it's not quite the same as current level where the exponential term is meaningful.  I hope the docstring makes this more clear. Open to suggestions. I recognize that you don't have the benefit of having the paper available yet, but the function 's usage should be clear without reference to the paper.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,268 @@
+    Parameters
+    ----------
+    v : numeric
+        Voltage at each point on the IV curve, from 0 to v_oc
```
Does this have to be a 1D array? How many points are required?

Does it have to also include the `i_sc`, `p_mp`, and `v_oc` points passed separately?

What about requiring user to simply pass an I-V curve that contains Isc, Pmp, and Voc.

cwhanse replied:
Added to the notes about number of points, and added the requirement for ordering to the docstring.

Functions should work with v and i as lists. I think the default approach is not to add type requirements until they are necessary.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,268 @@
+        Voltage at each point on the IV curve, from 0 to v_oc
+
+    i : numeric
+        Current at each point on the IV curve, from i_sc to 0
```
Does this have to be a 1D array? How many points are required?

Does it have to also include the `i_sc`, `p_mp`, and `v_oc` points passed separately?

What about requiring user to simply pass an I-V curve that contains Isc, Pmp, and Voc.

markcampanelli commented on pvlib/test/test_ivtools.py:
```diff
@@ -0,0 +1,88 @@
+def test_fit_cec_sam(cec_module_parameters):
+    sam_parameters = cec_module_parameters
+    cec_list_data = get_cec_params_cansol_cs5p_220p()
+    # convert from %/C to A/C and V/C
```
I don't like `%/degC` if it's actually `1/degC`.

cwhanse replied:
It's actually a mistake, holdover code from Matlab that should have been removed.  Corrected the expected values.

cwhanse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+        Short circuit current at STC
+    alpha_sc : float
+        Temperature coefficient of short circuit current at STC, A/C
+    beta_voc : float
```
`beta_voc` for internal consistency in pvlib. `alpha_sc` could be `alpha_isc` if we want to shift the pvlib name, but it's `alpha_sc` everywhere else where units are A/C, and that's an old very well-beaten horse.

I was incorrect about `beta_voc`.  The datasheet (or other source) value is used as input to the parameter extraction, it is not used as input when the IV curve is computed.  At the IV computation, beta_voc could be an output, once `Adjust` is known.  Its not an output currently, nor do I think it should be.

cwhanse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,262 @@
+    beta_voc : float
+        Temperature coefficient of open circuit voltage at STC, V/C
+    gamma_pmp : float
+        Temperature coefficient of power at maximum point point at STC, %/C
```
NREL's document doesn't name `gamma_pmp`, but the code uses `gamma_pmp` as the key to pass a value to the parameter solver. pvlib doesn't name `gamma_pmp` or `gamma_mp` - the only place a similar quantity is used is in the PVWatts function.  Interesting, but that's the current pvlib state.  If pvlib were to have a name I'd vote for `gamma_pmp`.

%/C according to the NREL code.

adriesse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,304 @@
+        R_s = np.nan
+        R_sh_ref = np.nan
+        Adjust = np.nan
+    return I_L_ref, I_o_ref, R_sh_ref, R_s, a_ref, Adjust
```
Is it really worth the effort to copy and rename the result?  The names don't get returned--only the values.

cwhanse replied:
Agree that the `return (tuple of variables)` is more compact and easier to read. I'd prefer to have the `if/then` structure with the single `return` in case we come up with a better fallback than returning `np.nan` when the SAM function doesn't work (e.g., SAM might expose some numerical options and we could retry the fitting).  There aren't options available now, that I know, but I like having the structure in place in case they become available.

FYI - `return [result[p] …` returns a list, `return result['Il'], result[…` returns a tuple.  They are a little different, I don't really understand why to prefer one over the other.  I see tuples used to pass values from functions, so I'm using one here.

mikofski replied:
AFAIK it's best practice to use a tuple as a keyword default input to a function because tuples are immutable,  whereas a list can change,  but I don't see why a list couldn't be the output.

From [Common Gotchas](https://docs.python-guide.org/writing/gotchas/):

>Seemingly the most common surprise new Python programmers encounter is Python’s treatment of mutable default arguments in function definitions.

markcampanelli replied:
I think sticking with a tuple as the return value is much more consistent with the current API.

I note the alternative "required keyword input arguments" and "dictionary of outputs" approach that I currently prefer for Python3 APIs: https://github.com/pvlib/pvlib-python/issues/660#issuecomment-464187557, which is much more forgiving w.r.t. ordering and is still convenient for call chaining, but is somewhat less performant (I think). Also remember that default values become a breakable part of the API.

wholmgren replied:
Agree that it is better to return tuples. I also like @adriesse's compact notation suggestion. You can do both if you wrap the list/generator comprehension in `tuple`: `tuple((result[p] for p in ['Il', 'Io', 'Rsh', 'Rs', 'a', 'Adj']))`. 

Does PySSC return a useful error message? If so, perhaps better to raise an exception to the user with the error message than return nan.

cwhanse replied:
PySSC prints messages to the screen. The message for failing to extract the parameters is not helpful. It looks to me that the [core functions](https://github.com/NREL/ssc/blob/1ee0dfdd9a46ce3ba80560bad650902c699f2780/shared/6par_solve.h) in the SSC check the solution for reasonable values of each parameter and for convergence to the data (look at the sanity() function if you are curious). An error code is set but no message is generated. Its a bit opaque in its current state.

cwhanse replied:
We could still set an exception if `nans` are being returned, but that would interrupt execution if someone was processing several input sets.

wholmgren replied:
> that would interrupt execution if someone was processing several input sets.

This is typically the recommended approach in Python. Let the user handle the exception unless you can safely assume what the right solution is for the vast majority of workflows. In the present state, the user would need to check that all results are not-nan to be sure that every input set succeeded. In this case, I'm ok with whatever the other reviewers of this PR prefer.

cwhanse replied:
The use case I have in mind is recreating the CEC model parameter database from the CEC's module characteristics data. There's 2000+ lines to process. An Exception that stops processing on the first line where the SAM algorithms failed to converge would be a bit frustrating, in my opinion.

@mikofski @adriesse what do you think?

adriesse replied:
I'm relying more an more on exceptions in my work and it doesn't seem like a bad choice here.  `try/except` will let you keep processing and let you choose whether to store nans, skip that module, log the error, or whatever.

markcampanelli replied:
While I think @cwhanse’s use case could still work with these exceptions, I personally like to reserve exceptions for truly exceptional cases, whereas here a numerical method failed “gracefully” (as opposed to exceptionally). That said, asking the user to search for nan’s to determine that things failed gracefully seems less than desirable.

If we go the Exception route, then I think we should define a custom pvlib exception that can be used throughout for this situation to unify the API. (Defining the scope “this situation” may be hard though!)

mikofski replied:
There was a similar discussion on the [vectorize newton PR in SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.newton.html), and they opted to display a `RuntimeWarning`, unless the optional arg `disp=True` in which case raise [`RuntimeError`](../../../scipy/scipy/blob/master/scipy/optimize/zeros.py#L361). However, the use of the keyword `disp` was considered awkward, and only maintained  for backwards compatibility. The [statsmodels author's opinion was a return flag](../../../scipy/scipy/pull/9568#issuecomment-444168546), similar to Cliff's NaNs idea. Someone else thought it should fall back to another algorithm. Etc. Lot's of competing opinions, no right answer.

My personal opinion is to favor exceptions. I think in MATLAB and other languages we learn to fear exceptions, but Python tends to embrace them. I think the argument for raising an exception is that it informs the user and let's them quickly decide how to handle it. It also avoids a lot of type-checking and null-pointer testing which isn't necessary in a memory-managed language. If I were running an example with 2000 points, I would want to know right away that it wasn't working, so I could fix it and move on. I think I would be more frustrated if the fitting continued without telling me anything, and then I ended up with an array full of NaNs. Aaaarrrrgggghhhh!

adriesse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,304 @@
+        IL = (1 + Gp * Rs) * beta0
+        # calculate I0
+        I0_v_mp = _calc_I0(IL, i_mp, v_mp, Gp, Rs, beta3)
+        I0_v_oc = _calc_I0(IL, 0, v_oc, Gp, Rs, beta3)
```
In keeping with the pattern of `beta_voc` would you prefer `I0_vmp` and `I0_voc`?

cwhanse replied:
yes

adriesse commented on pvlib/test/test_ivtools.py:
```diff
@@ -0,0 +1,90 @@
+            i_sc=cec_list_data['I_sc_ref'], alpha_sc=cec_list_data['alpha_sc'],
+            beta_voc=cec_list_data['beta_voc'],
+            gamma_pmp=cec_list_data['gamma_pmp'],
+            cells_in_series=cec_list_data['cells_in_series'], temp_ref=25)
```
I'm trying hard to conform to pvlib with naming in my own code, so looking more closely at this than necessary perhaps.  Do `V_mp_ref` etc. establish a new pattern?  I thought I had detected that lower case `v` was usually followed by an underscore, but not upper case `V`.

cwhanse replied:
lower case is preferred for variables. In this case, `V_mp_ref`, `I_sc_ref`, et al. are key values from the column headings in the CEC module parameter data, and we're stuck with them unless we rewrite the io function.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,304 @@
+        elif (I0_v_oc > 0):
+            I0 = I0_v_oc
+        else:
+            I0 = np.nan
```
In this case, should you set all the other parameters to be nan too?

Docstring states: "Returns ``NaN`` for each parameter if the fitting is not successful."

cwhanse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,304 @@
+        inclusive [V]
+
+    i : numeric
+        Current at each point on the IV curve, decreasing from i_sc to 0 [A]
```
Agree. My thought is that IV curve data cleaning should be handled elsewhere, so it can be used by other IV fitting functions.

adriesse replied:
Is it a strict requirement for the algorithm and/or implementation that voltage be decreasing and current increasing?

cwhanse replied:
Ah, I see the point now.  No it is not required that `v` increases and `i` decreases. It is reasonable to expect one of these two is sorted, but not necessarily the other.

Reconsidering having `v` and `i` as separate inputs: reaction to replacing `v, i` with a 2d ndarray?

markcampanelli replied:
IIRC, PVfit is designed to take arbitrarily ordered I-V curves, just so long as the I and V values correspond to each other in each vector. (Of course, my main design goal was to support Isc->Voc and Voc->Isc sweeps without the user having to pre-process the I-V curve.)

You might consider that here too. Use filters like V[V <= Vlim] and I[V <= Vlim] to filter out arbitrary V and I subvectors used to sequentially fit the different model parameters to different parts of the curve. Usually the underlying numpy/scipy regression routines don’t require any ordering.

I’m less inclined towards using 2D arrays.

cwhanse replied:
Sorted arrays are assumed where `numpy.searchsorted` is used. Filters could be used instead.

cwhanse commented:
> This is the last of my three-part review, covering the Sandia SDE fit.
> One general comment is that the interface to fit_sde_sandia() would be a bit more user friendly if only a "reasonably dense" I-V curve were necessary as a minimum input and all the other required information were derived automatically (perhaps allowing the user to provide overriding values).

Good idea to have, e.g., `v_oc=None` and if so, find the value in the I-V curve.

cwhanse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,304 @@
+
+    .. math::
+
+        I = IL - I0*exp((V+I*Rs)/(nNsVth)) - (V + I*Rs)/Rsh
```
Thanks for this. We can still edit the conference paper so I'm going to revise the presentation as you suggest. It is better justified this way.

cwhanse commented on pvlib/test/test_ivtools.py:
```diff
@@ -0,0 +1,90 @@
+            celltype='polySi', v_mp=0.45, i_mp=5.25, v_oc=0.55, i_sc=5.5,
+            alpha_sc=0.00275, beta_voc=0.00275, gamma_pmp=0.0055,
+            cells_in_series=1, temp_ref=25)
+    assert all(np.isnan([I_L_ref, I_o_ref, R_sh_ref, R_s, a_ref, Adjust]))
```
Did you divide `beta_voc` by the number of cells?

markcampanelli replied:
I tried it both ways, but I may have borked it. I will revisit and post the failing case.

markcampanelli replied:
I tried it both ways, but I may have borked it. I will revisit and post the failing case.

markcampanelli replied:
```python
# Copied from get_cec_params_cansol_cs5p_220p()
cec_list_data = {'V_mp_ref': 46.6, 'I_mp_ref': 4.73, 'V_oc_ref': 58.3, 'I_sc_ref': 5.05, 'alpha_sc': 0.0025,
                 'beta_voc': -0.19659, 'gamma_pmp': -0.43, 'cells_in_series': 96}
print(cec_list_data)
print(ivtools.fit_cec_sam(
    celltype='polySi', v_mp=cec_list_data['V_mp_ref'], i_mp=cec_list_data['I_mp_ref'],
    v_oc=cec_list_data['V_oc_ref'], i_sc=cec_list_data['I_sc_ref'], alpha_sc=cec_list_data['alpha_sc'],
    beta_voc=cec_list_data['beta_voc'], gamma_pmp=cec_list_data['gamma_pmp'],
    cells_in_series=cec_list_data['cells_in_series'], temp_ref=25))

# Normalize for a single cell of same area.
cec_list_data_one_cell = {'V_mp_ref': 46.6 / cec_list_data['cells_in_series'], 'I_mp_ref': 4.73,
                          'V_oc_ref': 58.3 / cec_list_data['cells_in_series'], 'I_sc_ref': 5.05,
                          'alpha_sc': 0.0025, 'beta_voc': -0.19659 / cec_list_data['cells_in_series'],
                          'gamma_pmp': -0.43, 'cells_in_series': 1}
print(cec_list_data_one_cell)
print(ivtools.fit_cec_sam(
    celltype='polySi', v_mp=cec_list_data_one_cell['V_mp_ref'], i_mp=cec_list_data_one_cell['I_mp_ref'],
    v_oc=cec_list_data_one_cell['V_oc_ref'], i_sc=cec_list_data_one_cell['I_sc_ref'],
    alpha_sc=cec_list_data_one_cell['alpha_sc'], beta_voc=cec_list_data_one_cell['beta_voc'],
    gamma_pmp=cec_list_data_one_cell['gamma_pmp'], cells_in_series=cec_list_data_one_cell['cells_in_series'],
    temp_ref=25))
 
I_L_ref, I_o_ref, R_sh_ref, R_s, a_ref, Adjust = ivtools.fit_cec_sam(
    celltype='polySi', v_mp=cec_list_data['V_mp_ref'], i_mp=cec_list_data['I_mp_ref'], v_oc=cec_list_data['V_oc_ref'],
    i_sc=cec_list_data['I_sc_ref'], alpha_sc=cec_list_data['alpha_sc'], beta_voc=cec_list_data['beta_voc'],
    gamma_pmp=cec_list_data['gamma_pmp'], cells_in_series=cec_list_data['cells_in_series'], temp_ref=25)
print(I_L_ref, I_o_ref, R_sh_ref, R_s, a_ref, Adjust)
```
gives
```terminal
{'V_mp_ref': 46.6, 'I_mp_ref': 4.73, 'V_oc_ref': 58.3, 'I_sc_ref': 5.05, 'alpha_sc': 0.0025, 'beta_voc': -0.19659, 'gamma_pmp': -0.43, 'cells_in_series': 96}
(5.056069368389881, 9.958811045933538e-11, 835.9279522043745, 1.00466417158272, 2.3663904878436925, 2.2270101674959184)
{'V_mp_ref': 0.48541666666666666, 'I_mp_ref': 4.73, 'V_oc_ref': 0.6072916666666667, 'I_sc_ref': 5.05, 'alpha_sc': 0.0025, 'beta_voc': -0.0020478124999999997, 'gamma_pmp': -0.43, 'cells_in_series': 1}
6parsolve simulation error
 : could not solve, check inputs
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-10-648913fa6eeb> in <module>
     20     alpha_sc=cec_list_data_one_cell['alpha_sc'], beta_voc=cec_list_data_one_cell['beta_voc'],
     21     gamma_pmp=cec_list_data_one_cell['gamma_pmp'], cells_in_series=cec_list_data_one_cell['cells_in_series'],
---> 22     temp_ref=25))
     23 
     24 I_L_ref, I_o_ref, R_sh_ref, R_s, a_ref, Adjust = ivtools.fit_cec_sam(

~/Documents/IMS/SolarPV/PVfit/pvlib-python/pvlib/ivtools.py in fit_cec_sam(celltype, v_mp, i_mp, v_oc, i_sc, alpha_sc, beta_voc, gamma_pmp, cells_in_series, temp_ref)
    102                       'Adj']])
    103     else:
--> 104         raise RuntimeError('Parameter estimation failed')
    105 
    106 

RuntimeError: Parameter estimation failed
```

cwhanse replied:
Thanks @markcampanelli for confirming that case as a failure for the SAM algorithm. I don't have the ability to run the SAM code in a debugging environment to see what criterion is causing the failure. The error condition is set by the function `sanity` [here](https://github.com/NREL/ssc/blob/1ee0dfdd9a46ce3ba80560bad650902c699f2780/shared/6par_solve.h) you can see what is being checked.  There's a numeric error code but it's not passed out, only the fact of an error. @nickdiorio @janinefreeman

janinefreeman replied:
Hi @cwhanse @markcampanelli - confirmed that the example case posted does fail in the solver code. We do occasionally run into convergence errors for various input cases. I won't have time to take a look at it for a while, but Mark, if you care to dig in while you're still fresh on the topic, we have instructions for debugging the SSC code both from Python (https://github.com/NREL/ssc/wiki/Debug-SSC-in-Windows-using-Python,-VS-2017,-and-SAM-SDK) and from the SDKtool (https://github.com/NREL/ssc/wiki/Debug-SSC-in-Windows-using-VS-2017-and-SDKtool)

cwhanse commented:
@wholmgren could you look at the travis-ci failure for py3.5-min?  Something's changed outside of pvlib, I don't know what to look for.

wholmgren commented:
Ah, frustrating. Might need to install those package versions using pip instead of conda. Or we can bump up the min supported versions.

markcampanelli commented:
I think this line still needs to be added to `pvlib/__init__.py`:
```python
from pvlib import ivtools
```
cc @wholmgren

cwhanse commented:
Ready for merge? @mikofski @markcampanelli @adriesse @wholmgren

mikofski commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,334 @@
+    beta0 = np.nan
+    beta1 = np.nan
+    idx = np.searchsorted(v, vlim * v_oc)
+    while idx <= len(v):
```
- Redundant `len(v)` calculated each loop
- suggest either instead define `numpoints = len(v)` on line before loop
- or even better just use `try: except IndexError` might perform best?

```python
    try:
        coef = np.polyfit(v[:idx], i[:idx], deg=1)
        if coef[0] < 0:
            # intercept term
            beta0 = coef[1].item()
            # sign change of slope to get positive parameter value
            beta1 = -coef[0].item()
            break
        else:
            idx += 1
    except IndexError as e:
        raise RuntimeError("no fit")

mikofski replied:
Also, since you use `break`, a good pattern might be a for loop

    for idx in len(v):
        coef = np.polyfit(v[:idx], i[:idx], deg=1)
        if coef[0] < 0:
            # intercept term
            beta0 = coef[1].item()
            # sign change of slope to get positive parameter value
            beta1 = -coef[0].item()
            break

mikofski commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,334 @@
+    x = np.array([np.ones_like(v), v, i]).T
+    # Select points where y > ilim * i_sc to regress log(y) onto x
+    result = np.linalg.lstsq(x[y > ilim * i_sc], np.log(y[y > ilim * i_sc]),
+                             rcond=None)
```
Calculate desired indices, `y > ilim * i_sc` only once, eg, `i_nonlin = (y > ilim * i_sc)` then use in `x[i_nonlin]`

mikofski commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,334 @@
+def _calculate_sde_parameters(beta0, beta1, beta3, beta4, v_mp, i_mp, v_oc):
+    failed = False
+    if any(np.isnan([beta0, beta1, beta3, beta4])):
+        failed = True
```
Zen of python says explicit is better.

Here and below, and then above in the calling function, imo use exceptions and propagate them up.

- remove `failed = False`
- instead of `failed = False`, `raise RuntimeError("beta coefficient was NaN")` 
- remove `else` and dedent, because next block can't be reached if no error
- on line 322, check `I0`
       if (I0_vmp > 0) and (I0_voc > 0):
           raise RuntimeError("I0 was less than zero")
- remove `if failed: results = None` and remove `else`, then just `return IL, I0, Rsh, Rs, nNsVth`
- finally in `fit_sde_sandia` on line 240 instead of `if result is None:` either just remove that section and do nothing, because you are already raising an exception, or place a try above and put except there instead, to modify the message.

cwhanse replied:
I'm all for improving the code for others to understand. One challenge here is that `np.linalg.lstsq` and `np.polyfit` throw an unhelpful "does not converge" error if there is any NaN in the data.  I want to avoid exiting with that error while assigning np.nan to SDE parameters that aren't determined.

cwhanse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,334 @@
+
+    Notes
+    -----
+    Inputs ``v``, ``i``, ``v_mp``, ``v_oc``, ``i_mp`` and ``i_sc`` are assumed
```
Thanks for the catch

cwhanse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,334 @@
+
+    .. math::
+
+        I = IL - I0*exp((V+I*Rs)/(nNsVth) - 1) - (V + I*Rs)/Rsh
```
Cut and paste mistake, thanks

mikofski commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,327 @@
+        raise RuntimeError("Parameter extraction failed: I0 is negative. Try "
+                           "increasing the number of data points in the IV "
+                           "curve")
+        I0 = np.nan
```
I believe this line will never be used

cwhanse replied:
Ah right. It's not a warning.

cwhanse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,322 @@
+        raise RuntimeError("Parameter extraction failed",
+                           "  beta0: {}, beta1: {}, beta3: {}, beta4: {}"
+                           .format(beta0, beta1, beta3, beta4))
+        return (np.nan, np.nan, np.nan, np.nan)
```
Delete this `return`

cwhanse commented:
> Did you mean to delete this line?
> elif (I0_vmp <= 0) and (I0_voc <= 0):
>     I0 = np.nan

Yes. Extraction fails if I0 is NaN (because other parameters are NaN) or if I0 < 0.

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -35,17 +35,12 @@ def fit_cec_sam(celltype, v_mp, i_mp, v_oc, i_sc, alpha_sc, beta_voc,
     cells_in_series : int
         Number of cells in series
     temp_ref : float, default 25C
```
Nit: Now that the units are elsewhere, perhaps this should just be 25.

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,322 @@
+    # Select points where y > ilim * i_sc to regress log(y) onto x
+    idx = (y > ilim * i_sc)
+    result = np.linalg.lstsq(x[idx], np.log(y[idx]),
+                             rcond=None)
```
looks like this would fit on one line

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,322 @@
+    I0_vmp = _calc_I0(IL, i_mp, v_mp, Gp, Rs, beta3)
+    I0_voc = _calc_I0(IL, 0, v_oc, Gp, Rs, beta3)
+    if any(np.isnan([I0_vmp, I0_voc])) or ((I0_vmp <= 0) and (I0_voc <= 0)):
+        raise RuntimeError("Parameter extraction failed: I0 is undetermined. ")
```
extra space

cwhanse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,322 @@
+    :py:func:`pvsystem.singlediode` for definition of the parameters.
+
+    The extraction method [2] proceeds in four steps:
+        1) In the single diode equation, replace Rsh = 1/Gp and re-arrange
```
It doesn't render correct.  Do you know of a summary of latex commands available in rst? what I know in latex, doesn't work.

mikofski replied:
In docstring I often have to escape latex twice or preface the entire for string with raw: `r"""\beta"""` or `"""\\beta"""`

markcampanelli commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,349 @@
+
+
+def _calc_I0(IL, I, V, Gp, Rs, beta3):
+    return (IL - I - Gp * V - Gp * Rs * I) / np.exp(beta3 * (V + Rs * I))
```
Nit: This might be more immediately understood in terms of `nNsVth` instead of `beta3` = 1 / `nNsVth`.

wholmgren commented:
@cwhanse try putting the `r` at very beginning of the docstring:

```python
def my_function():
    r"""
    the docstring with \latex
    """
```

no promises...

cwhanse commented:
> @cwhanse try putting the r at very beginning of the docstring:
> def my_function():
>     r"""
>     the docstring with \latex
>     """
> no promises...

Worth a try. I don't understand why [this](https://github.com/pvlib/pvlib-python/blob/e3635c2faa4ef990cbbae6cf4432d9773d00f31f/pvlib/pvsystem.py#L2840) works but this docstring is garbled.

adriesse commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,351 @@
+    x = np.array([np.ones_like(voltage), voltage, current]).T
+    # Select points where y > ilim * i_sc to regress log(y) onto x
+    idx = (y > ilim * i_sc)
+    result = np.linalg.lstsq(x[idx], np.log(y[idx]), rcond=None)
```
Out of curiosity, why `lstsq` here and `polyfit` above?

cwhanse replied:
`polyfit` is single variable only. Here we need multiple regression.

adriesse replied:
Oh yes, fairly obvious now that you mention it.

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,349 @@
+def fit_sde_sandia(voltage, current, v_oc=None, i_sc=None, v_mp_i_mp=None,
+                   vlim=0.2, ilim=0.1):
+    """
+    Fits the single diode equation to an IV curve.
```
alternative function name: `fit_single_diode_sandia`. Not a blocker.

cwhanse replied:
better name

cwhanse replied:
and rethinking the name change, `sde` is more appealing because we anticipate `sdm` in this PR, perhaps `sde_recomb`, `dde` and `ddm` in the future.

cwhanse commented on pvlib/test/test_ivtools.py:
```diff
@@ -0,0 +1,105 @@
+@requires_pysam
+def test_fit_cec_sam(cec_module_parameters, get_cec_params_cansol_cs5p_220p):
+    sam_parameters = cec_module_parameters
+    cec_list_data = get_cec_params_cansol_cs5p_220p
```
I tried that originally, but `get_cec_params_cansol_cs5p_220p['gamma_pmp']` was giving a KeyError if I remember correctly. Renaming got around the error. I'm sure I don't fully understand how the test fixtures work.

cwhanse commented:
> @cwhanse try putting the r at very beginning of the docstring:
> def my_function():
> r"""
> the docstring with \latex
> """
> no promises...
> 
> Worth a try. May be why this works but the docstring in this PR is garbled.

Docstrings render OK [readthedocs](https://cwhanse-pvlib-python.readthedocs.io/en/ivtools/api.html#functions-for-fitting-pv-models).  I'd prefer if the `Returns` list items lined up with the bullets, but its acceptable, I think. 

I think I have addressed all comments.

wholmgren commented on docs/sphinx/source/whatsnew/v0.7.0.rst:
```diff
@@ -19,9 +19,31 @@ Testing
 * Mark Campanellli (:ghuser:`markcampanelli`)
 * Will Holmgren (:ghuser:`wholmgren`)
+* Cliff Hansen (:ghuser:`cwhanse`)
+
```
blank line

wholmgren commented on docs/sphinx/source/api.rst:
```diff
@@ -282,6 +282,14 @@ PVsyst model
+   :toctree: generated/
+
+    ivtools.fit_single_diode_sandia
+    ivtools.fit_cec_sam
```
Maybe `fit_single_diode_cec_sam` to more closely mimic the sandia function?

markcampanelli replied:
I kinda like the single-diode equation (SDE) vs. single-diode model (SDM) distinction, and I don’t mind the abbreviations if they are expanded in the docstring. This leads to something succinct and descriptive like:

`fit_sde_sandia`
`fit_sdm_cec_sam`

`action_thing_[variant_]source`

I don’t know if you want to add a variant to the SDE (to cover potential variants, such as with a reverse breakdown component). Also, the variant and source could repeat in some cases, leading to some stuttering.

cwhanse replied:
@markcampanelli good suggestion for naming patterns. There are two different applications here: fitting the SD equation to an IV curve, and fitting a SD model (a set of equations) to some data (in the case of the CEC model and the SAM method, datasheet information). I changed `sde` to `single_diode` in one function name, but I don't care for `single_diode_equation` and `single_diode_model` in the names.

I'm going to open a new issue to extend the single diode documentation to lay out the terminology: single diode equation, single diode model, 5 parameters vs. model parameters.

wholmgren replied:
Works for me.

wholmgren commented on pvlib/ivtools.py:
```diff
@@ -0,0 +1,350 @@
+
+    Raises
+    ------
+        RuntimeError if parameter extraction is not successful.
```
Is this indent necessary?

cwhanse replied:
nope

wholmgren commented:
The coverage report indicates that a handful of the error handling `if/elif` blocks are not tested (click on the "Diff" tab of the codecov/patch link). We should add tests that hit those blocks. We can ignore the missed coverage in `conftest.py` and the `ImportError`.

wholmgren commented on pvlib/test/test_ivtools.py:
```diff
@@ -67,6 +67,16 @@ def test_fit_sde_sandia(get_test_iv_params):
     result = ivtools.fit_sde_sandia(voltage=testcurve['v'],
                                     current=testcurve['i'], vlim=0.1)
     assert np.allclose(result, expected, rtol=5e-5)
+    # bad IV curves for coverage of if/then in _calculate_sde_parameters
```
at the risk of nitpicking, this looks like it should be its own test:

```python
def test_fit_sde_sandia_bad_iv(get_bad_iv_curves):
```

wholmgren commented:
@cwhanse from a quick glance it seems that the only remaining issues are cleaning up the whatsnew file (duplicates introduced in merge) and fixing a few tests. Let me know if there's anything else to double check, otherwise let's consider this complete.

wholmgren commented:
@cwhanse can you make a new pr to clean up the whats new file?
